File: backend/app/models/product_batch.py
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ProductBatch(Base):
    __tablename__ = "product_batches"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    product_id: Mapped[int] = mapped_column(
        ForeignKey("products.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    batch_number: Mapped[str | None] = mapped_column(String(80), index=True)
    expiration_date: Mapped[date | None] = mapped_column(Date, index=True)
    quantity: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False, default=0)
    unit: Mapped[str] = mapped_column(String(20), nullable=False, default="un")
    source_type: Mapped[str | None] = mapped_column(String(40), index=True)
    source_id: Mapped[int | None] = mapped_column(index=True)
    source_number: Mapped[str | None] = mapped_column(String(60), index=True)
    supplier_name: Mapped[str | None] = mapped_column(String(180))
    invoice_number: Mapped[str | None] = mapped_column(String(30))
    invoice_series: Mapped[str | None] = mapped_column(String(20))
    notes: Mapped[str | None] = mapped_column(Text)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    product = relationship("Product", back_populates="batches")

File: admin_app/admin_flutter/lib/models/product_batch.dart
class ProductBatch {
  const ProductBatch({
    required this.id,
    required this.productId,
    required this.quantity,
    required this.unit,
    required this.active,
    required this.createdAt,
    required this.updatedAt,
    this.batchNumber,
    this.expirationDate,
    this.sourceType,
    this.sourceId,
    this.sourceNumber,
    this.supplierName,
    this.invoiceNumber,
    this.invoiceSeries,
    this.notes,
  });

  final int id;
  final int productId;
  final String? batchNumber;
  final String? expirationDate;
  final double quantity;
  final String unit;
  final String? sourceType;
  final int? sourceId;
  final String? sourceNumber;
  final String? supplierName;
  final String? invoiceNumber;
  final String? invoiceSeries;
  final String? notes;
  final bool active;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory ProductBatch.fromJson(Map<String, dynamic> json) {
    return ProductBatch(
      id: json['id'] as int,
      productId: json['product_id'] as int,
      batchNumber: json['batch_number'] as String?,
      expirationDate: json['expiration_date'] as String?,
      quantity: _toDouble(json['quantity']),
      unit: json['unit'] as String? ?? 'un',
      sourceType: json['source_type'] as String?,
      sourceId: json['source_id'] as int?,
      sourceNumber: json['source_number'] as String?,
      supplierName: json['supplier_name'] as String?,
      invoiceNumber: json['invoice_number'] as String?,
      invoiceSeries: json['invoice_series'] as String?,
      notes: json['notes'] as String?,
      active: json['active'] as bool? ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  static double _toDouble(Object? value) {
    if (value == null) return 0;
    if (value is num) return value.toDouble();
    return double.tryParse(value.toString()) ?? 0;
  }
}
